File: android/src/main/java/io/mosip/tuvali/rnModule/RNWalletModule.kt
package io.mosip.tuvali.rnModule

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import io.mosip.tuvali.wallet.IWallet

class RNWalletModule(
  private val eventEmitter: RNEventEmitter,
  private val wallet: IWallet,
  reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext) {

  init {
    wallet.subscribe {
      eventEmitter.emitEvent(RNEventMapper.toMap(it))
    }
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun startConnection(uri: String) {
    wallet.startConnection(uri)
  }

  @ReactMethod
  fun sendData(payload: String) {
    wallet.sendData(payload)
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun disconnect() {
    wallet.disconnect()
  }

  override fun getName(): String {
    return NAME
  }

  companion object {
    const val NAME = "WalletModule"
  }
}

File: ios/RNModules/RNWalletModule.swift
import Foundation

@available(iOS 13.0, *)

class WalletModule: RCTEventEmitter {
    var wallet: WalletProtocol = Wallet()
    var tuvaliVersion: String = "unknown"

    override init() {
        super.init()
        RNEventEmitter.sharedInstance.registerEventEmitter(producer: self)
        wallet.subscribe { event in
            RNEventEmitter.sharedInstance.emitEvent(eventMap: RNEventMapper.toMap(event))
        }
    }

    @objc func noop() -> Void {}

    @objc func startConnection(_ uri: String) {
        wallet.startConnection(uri)
    }

    @objc func disconnect(){
        wallet.disconnect()
    }

    @objc func sendData(_ payload: String) {
        wallet.send(payload)
    }

    @objc override func supportedEvents() -> [String]! {
        return RNEventEmitter.sharedInstance.allEvents
    }

    @objc override static func requiresMainQueueSetup() -> Bool {
        return false
    }
}
